File: leetcode-go/235.go
package main

/**
 * Definition for a binary tree node.
 *
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func searchAncestor(root, p, q *TreeNode, ancestorFound *bool) *TreeNode {
	if root == nil || *ancestorFound {
		return nil
	}
	foundNode := false
	if root == p || root == q {
		foundNode = true
	}
	inLeft := searchAncestor(root.Left, p, q, ancestorFound)
	inRight := searchAncestor(root.Right, p, q, ancestorFound)
	checkIfCommonAncestor := (foundNode && (inLeft != nil || inRight != nil)) || (inLeft != nil && inRight != nil)
	if !(*ancestorFound) && checkIfCommonAncestor {
		*ancestorFound = true
		return root
	}
	if foundNode {
		return root
	}
	if inLeft != nil {
		return inLeft
	}
	if inRight != nil {
		return inRight
	}
	return nil
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	ancestorFound := false
	ancestor := searchAncestor(root, p, q, &ancestorFound)
	// fmt.Println(ancestor)
	return ancestor
}

File: leetcode-js/235.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
    let result = null;

    const treeTraverse = (node) => {
        if (!node || result !== null) return [false, false];
        let res = [false, false];
        let [pLeft, qLeft] = treeTraverse(node.left);
        let [pRight, qRight] = treeTraverse(node.right);
        res[0] = (res[0] || pLeft || pRight);
        res[1] = (res[1] || qLeft || qRight);
        if (node === p) res[0] = true;
        if (node === q) res[1] = true;
        if (res[0] && res[1]) {
            result = node;
            return [false, false];
        }
        return res;
    }
    treeTraverse(root);
    return result;
};